### PR TITLE
dasSpirv Phase 10.3: fragment realism rails (discard/derivatives/textureSize/gl_FragDepth)

### PR DESCRIPTION
**Pull request by borisbat**
Phase 10.3 of the dasSpirv roadmap — the fragment-stage foundation rails a real fragment shader needs (alpha-test discard, edge-aware AA via derivatives, mip/texel math via size queries, depth replacement). Stacks on 10.1 (#3159) + 10.2 (#3160), both merged. Five new census opcodes; each rail is its own intrinsic in `spirv_builtins.das`, lowered in the emitter's call dispatch.

## Rails

- **`discard()` → `OpKill`** — a block terminator: sets `ctx.terminated`, so a statement following it on the same path is rejected as unreachable (the existing dead-code guard). Matches glslang lowering `discard` → `OpKill`. Fragment-only.
- **`dFdx`/`dFdy`/`fwidth` → `OpDPdx`/`OpDPdy`/`OpFwidth`** — screen-space derivatives over the 2×2 quad. Fragment-only; the basic forms need only the `Shader` capability. Concrete per-width overloads (float/float2/3/4), not a generic — see finding 1.
- **`textureSize(sampler, lod)` → `OpImage` + `OpImageQuerySizeLod`** — pulls in the `ImageQuery` capability lazily (deduped via a `ctx` flag). int2 for 2D/Cube, int3 for 3D / 2D-array. Stage-agnostic.
- **`gl_FragDepth` write → BuiltIn `FragDepth` Output + `DepthReplacing` execution mode** — the exec mode is emitted only when the builtin is referenced (`classify_global` sets `ctx.uses_frag_depth`, read after dependency collection).

## Findings

1. **Generic builtins get mangled instance call-names** (`spirv_builtins`dFdx`5408…`), so the emitter's `name == "dFdx"` dispatch can't match a `def dFdx(p : auto(TT))`. Every existing builtin is a concrete overload for exactly this reason; the derivatives follow suit (4 widths × 3 functions).
2. **`OpKill` is a terminator, so `discard()` reuses the existing terminated-block machinery for free** — no new control-flow plumbing.

## Testing

`fragx` fixture + `test_fragment` gate (each opcode/decoration asserted positionally + spirv-val). Census extended to `phase10_3_emitter_opcodes` (= 10.2 set + the 5 new opcodes), unioned over the new fixture. **interp + JIT 96/96**, spirv-val clean on every blob, lint + format clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tests/spirv/_spirv_common.das [hunk 1]
```diff
@@ -756,6 +756,32 @@ def public gswizzle_words : array<uint> {
     return <- w
 }
 
+// ===== Phase-10.3 fragment realism: discard (OpKill), screen-space derivatives (OpDPdx/OpDPdy/
+// OpFwidth), textureSize (OpImageQuerySizeLod + the ImageQuery capability), and a gl_FragDepth write
+// (BuiltIn FragDepth Output -> the DepthReplacing execution mode). All fragment-only. discard() sits
+// in an if-branch (it is a block terminator); dFdx/dFdy take the whole float2 global, fwidth a scalar. =====
```
[the remark this answers is not in the document]
Comment (borisbat) [reply tests/spirv/_spirv_common.das]: Correct — fixed in 440b6f5e5. The header now scopes "fragment-only" to discard/derivatives/gl_FragDepth and notes textureSize is stage-agnostic (exercised here in a fragment shader for convenience).

File: modules/dasSpirv/spirv/spirv_emit.das [hunk 2]
```diff
@@ -56,6 +56,8 @@ struct EmitCtx {
     terminated    : bool                        // current block already has a terminator
     glsl_ext      : uint                        // OpExtInstImport "GLSL.std.450" set id (0 = not imported yet)
     stage         : ShaderStage                 // the shader stage (texture() implicit-LOD sampling is fragment-only)
+    uses_frag_depth : bool                      // gl_FragDepth is written -> emit the DepthReplacing execution mode
+    image_query_cap : bool                      // OpCapability ImageQuery already emitted (textureSize)
```
[the remark this answers is not in the document]
Comment (borisbat) [reply modules/dasSpirv/spirv/spirv_emit.das:60]: Fixed in 440b6f5e5. The flag is indeed set at classification (on reference). Since gl_FragDepth is Output-only, a reference is necessarily a write, so the behavior is correct — the comment now says so explicitly rather than implying a write-specific check.

File: modules/dasSpirv/spirv/spirv_emit.das [hunk 3]
```diff
@@ -266,6 +269,10 @@ def private classify_global(var m : SpirvModule; var ctx : EmitCtx; v : Variable
         emit(m, SEC_DECOR, SpvOp.Decorate, vid, uint(SpvDecoration.BuiltIn), uint(bi.bi))
         ctx.globals |> insert(intptr(v), GlobalInfo(var_id = vid, storage = bi.storage, value_type = vt, is_ssbo = false))
         ctx.interface_ids |> push(vid)   // Input + Output builtins both list in the entry interface
+        // Writing gl_FragDepth replaces the fixed-function depth -> the shader must declare DepthReplacing.
+        if (bi.bi == SpvBuiltIn.FragDepth) {
+            ctx.uses_frag_depth = true
+        }
```
[the remark this answers is not in the document]
Comment (borisbat) [reply modules/dasSpirv/spirv/spirv_emit.das]: Fixed in 440b6f5e5 — the comment now says the flag is set when gl_FragDepth is referenced at classification, and notes that since it's Output-only a reference is a write.

File: modules/dasSpirv/spirv/spirv_emit.das [hunk 4]
```diff
@@ -2150,6 +2233,10 @@ def generate_spirv(fn : FunctionPtr; var errors : array<string>; stage : ShaderS
         delete em
     } elif (stage == ShaderStage.Fragment) {
         emit(m, SEC_EXECMODE, SpvOp.ExecutionMode, id_main, uint(SpvExecutionMode.OriginUpperLeft))
+        // Writing gl_FragDepth (detected by classify_global above) replaces the fixed-function depth.
+        if (ctx.uses_frag_depth) {
+            emit(m, SEC_EXECMODE, SpvOp.ExecutionMode, id_main, uint(SpvExecutionMode.DepthReplacing))
+        }
```
[the remark this answers is not in the document]
Comment (borisbat) [reply modules/dasSpirv/spirv/spirv_emit.das:2241]: Fixed in 440b6f5e5 — reworded to "gl_FragDepth referenced (detected by classify_global above; Output-only, so a reference is a write)".
